File: leetcode/218.天际线问题.py
from sortedcontainers import SortedList

class Node:
    def __init__(self, l = 0, r = 0, mx = 0):
        self.l = l
        self.r = r
        self.mx = mx
class SegmentTree:
    def __init__(self, n):
        self.f = [Node() for i in range(4 * n)]
    
    def pushup(self, u):
        self.f[u].mx = max(self.f[u * 2].mx, self.f[u * 2 + 1].mx)
    
    def pushdown(self, u):
        if self.f[u].mx:
            self.f[u * 2].mx = max(self.f[u * 2].mx, self.f[u].mx)
            self.f[u * 2 + 1].mx = max(self.f[u * 2 + 1].mx, self.f[u].mx)
            self.f[u].mx = 0
    def build(self, u, l, r):
        self.f[u].l, self.f[u].r = l, r
        if l >= r: 
            return 
        mid = (self.f[u].l + self.f[u].r) // 2
        self.build(2 * u, l, mid)
        self.build(2 * u + 1, mid + 1, r) 
        self.pushup(u)
    
    def modify(self, u, l, r, d):
        if self.f[u].l >= l and self.f[u].r <= r:
            self.f[u].mx = max(self.f[u].mx, d)
            return
        self.pushdown(u)
        mid = (self.f[u].l + self.f[u].r) // 2
        if l <= mid:
            self.modify(u * 2, l, r, d)
        if r > mid:
            self.modify(u * 2 + 1, l, r, d)
        # self.pushup(u)
    
    def query(self, u, l, r):
        if self.f[u].l >= l and self.f[u].r <= r:
            return self.f[u].mx
        self.pushdown(u)
        mid = (self.f[u].l + self.f[u].r) // 2
        v = -inf
        if l <= mid: 
            v = self.query(2 * u, l, min(r, mid))
        if mid < r: 
            v = max(v, self.query(2 * u + 1, max(mid + 1, l), r))
        return v
class Solution:
    # 线段树解法, 维护区间最大值
    def getSkyline(self, buildings: List[List[int]]) -> List[List[int]]:
        # 离散化排序去重
        arr = []
        for x, y, _ in buildings:
            arr.append(x)
            arr.append(y)
        arr = sorted(set(arr))
        print(arr)
        n = len(arr)
        sg = SegmentTree(n + 1)
        sg.build(1, 1, n)
        for i, [x, y, h] in enumerate(buildings):
            l, r = bisect_left(arr, x) + 1, bisect_left(arr, y)
            sg.modify(1, l, r, h) # 区间修改最大值
        maxh = -1
        ans = []
        for i in range(n):
            x = i + 1
            t = sg.query(1, x, x)
            if t != maxh:
                ans.append([arr[i], t])
                maxh = t
        return ans        

    # 平衡树解法
    def getSkyline1(self, buildings: List[List[int]]) -> List[List[int]]:
        def cmp(a, b):
            if a[0] != b[0]:
                return a[0] - b[0]
            if a[0] == b[0]:
                if a[1] != b[1]:
                    return a[1] - b[1]
                elif a[1] == 0:
                    return b[2] - a[2]
                elif a[1] == 1:
                    return a[2] - b[2]
        arr = []
        for i, [l, r, h] in enumerate(buildings):
            arr.append((l, 0, h))
            arr.append((r, 1, h))
        arr.sort(key = cmp_to_key(cmp))
        st = SortedList()
        ans = []
        maxh = -1
        for x, lr, h in arr:
            if lr == 0:
                st.add(h)        
                if st[-1] > maxh:
                    ans.append([x, h])
                    maxh = st[-1]
            else:
                st.remove(h)
                if len(st) == 0:
                    ans.append([x, 0])
                    maxh = 0
                if len(st) and maxh > st[-1]:
                    ans.append([x, st[-1]])
                    maxh = st[-1]
        return ans

File: leetcode/218.天际线问题.cpp
const int N = 20010;
class Solution {
public:
    unordered_map<int, int> mp;
    typedef long long LL;
    struct Node{
        int l, r;
        int h; // h维护的是每一个区间的最大高度
    }tr[N * 4];

    void build(int p, int l, int r){
        tr[p].l = l, tr[p].r = r;
        if (l == r) return;
        int mid = (l + r) >> 1;
        build(p * 2, l, mid);
        build(p * 2 + 1, mid + 1, r);
    }
    void spread(int p){ //懒标记, 如果存在则向下更新儿子节点的最大高度
        if (tr[p].h){
            tr[p * 2].h = max(tr[p * 2].h, tr[p].h);
            tr[p * 2 + 1].h = max(tr[p * 2 + 1].h, tr[p].h);
            tr[p].h = 0;
        }
    }
    void modfiy(int p, int l, int r, int d){
        if (l <= tr[p].l && r >= tr[p].r){
            tr[p].h = max(tr[p].h, d);
            return;
        }
        spread(p);
        int mid = (tr[p].l + tr[p].r) >> 1;
        if (l <= mid) modfiy(p * 2, l, r, d);
        if (r >= mid + 1) modfiy(p * 2 + 1, l, r, d);
    }

    int query(int p, int x){ // 单点查询离散化的每一个点
        if (tr[p].l == tr[p].r){
            return tr[p].h;
        }
        spread(p);
        int mid = (tr[p].l + tr[p].r) >> 1;
        if (x <= mid) return query(p * 2, x);
        else return query(p * 2 + 1, x);
    }
    vector<vector<int>> getSkyline(vector<vector<int>>& bs) {
        vector<int> ans;
        vector<vector<int>> res;
        for (int i = 0; i < bs.size(); i ++){
            ans.push_back(bs[i][0]);
            ans.push_back(bs[i][1]);
        }
        //排序去重+离散化处理；
        sort(ans.begin(), ans.end());
        ans.erase(unique(ans.begin(), ans.end()), ans.end());
        int n = ans.size();
        for (int i = 0; i < n; i ++) {
            mp[ans[i]] = i + 1;
        }
        build(1, 1, n);
        for (int i = 0; i < bs.size(); i ++){
            int l = mp[bs[i][0]], r = mp[bs[i][1]] - 1;
            modfiy(1, l, r, bs[i][2]);
        }
        int maxv = 0;
        for (int i = 0; i < n; i ++){
            int x = query(1, mp[ans[i]]); 
            if (x != maxv) { // 判断与上一个节点是否相同。
                res.push_back({ans[i], x});
                maxv = x;
            }
        }   
        return res;
    }
};
